### PR TITLE
refactor: moving consul access and role interface

### PR DESCRIPTION
**Pull request by ItsRico**
Moving configureConsulAccess and createRole to go-mod-secrets to consolidate consul API.

Closes: #3227

Signed-off-by: Rico Chavez-Lopez <rchavezlopez@ucdavis.edu>

If your build fails due to your commit message not passing the build checks, please review the guidelines here: https://github.com/edgexfoundry/go-mod-secrets/blob/main/.github/Contributing.md

## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] I am not introducing a breaking change (if you are, flag in conventional commit message with `BREAKING CHANGE:` describing the break)
- [ x] I am not introducing a new dependency (add notes below if you are)
- [ x] I have added unit tests for the new feature or bug fix (if not, why?)
- [x ] I have fully tested (add details below) this the new feature or bug fix (if not, why?)
- [ ] I have opened a PR for the related docs change (if not, why?)
  <link to docs PR>

## Testing Instructions

Please refer to https://github.com/edgexfoundry/edgex-go/pull/4193

### CODE REVIEW

**Review by bnevis-i — DISMISSED** [review]
LGTM.  Merging hold until have corrected edgex-go PR to test against.

**Review by jim-wang-intel — DISMISSED** [review]

File: pkg/types/registryroles.go [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ *******************************************************************************/
+
+package types
+
+import "strings"
+
+type RegistryTokenType string
```
Comment (lenny-intel) [inline pkg/types/registryroles.go:20]: Chhange all use of term `registry` to `Consul` in code and comments
```suggestion
type ConsulTokenType string
```

File: pkg/types/registryroles.go [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+	 */
+	// ManagementType is the type of registry role can be used to create tokens when role-based API /consul/creds is called
+	// the management type of created tokens is automatically granted the built-in global management policy
+	ManagementType RegistryTokenType = "management"
+	// ClientType is the type of registry role that can be used to create tokens when role-based API /consul/creds is called
+	// the regular client type of created tokens is associated with custom policies
+	ClientType RegistryTokenType = "client"
+)
+
+type RegistryRole struct {
```
Comment (lenny-intel) [inline pkg/types/registryroles.go:36]:
```suggestion
type ConsulRole struct {
```

File: pkg/types/registryroles.go [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+	Local       bool     `json:"local,omitempty"`
+	TimeToLive  string   `json:"TTL,omitempty"`
+}
+
+type Policy struct {
+	ID   string `json:"ID"`
+	Name string `json:"Name"`
+}
+
+func NewRegistryRole(name string, tokenType RegistryTokenType, policies []Policy, localUse bool) RegistryRole {
```
Comment (lenny-intel) [inline pkg/types/registryroles.go:49]:
```suggestion
func NewConsulRole(name string, tokenType ConsulTokenType, policies []Policy, localUse bool) ConsulRole {
```

File: pkg/types/registryroles.go [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+	// as the policy name needs to be unique per API's requirement
+	policyNames := make([]string, 0, len(policies))
+	tempMap := make(map[string]bool)
+	for _, policy := range policies {
+		if _, exists := tempMap[policy.Name]; !exists {
+			policyNames = append(policyNames, policy.Name)
+		}
+	}
+
+	return RegistryRole{
```
Comment (lenny-intel) [inline pkg/types/registryroles.go:61]:
```suggestion
	return ConsulRole{
```

File: pkg/types/registryroles.go [hunk 5]
```diff
@@ -0,0 +1,69 @@
+/*******************************************************************************
```
Comment (lenny-intel) [inline pkg/types/registryroles.go:1]: change file name to consulroles.go

File: internal/pkg/vault/management.go [hunk 6]
```diff
@@ -193,3 +193,57 @@ func (c *Client) CheckSecretEngineInstalled(token string, mountPoint string, eng
...
+	})
+
+	return err
+}
+
+// ConfigureConsulAccess is to enable the Consul config access to the SecretStore via consul/config/access API
+// see the reference: https://www.vaultproject.io/api-docs/secret/consul#configure-access
+func (c *Client) ConfigureConsulAccess(secretStoreToken string, bootstrapACLToken string, registryHost string, registryPort int) error {
+	type ConfigAccess struct {
+		RegistryAddress   string `json:"address"`
```
Comment (lenny-intel) [inline internal/pkg/vault/management.go:229]: Here an others below.

```suggestion
		ConsulAddress   string `json:"address"`
```

File: internal/pkg/vault/management.go [hunk 7]
```diff
@@ -193,3 +193,57 @@ func (c *Client) CheckSecretEngineInstalled(token string, mountPoint string, eng
...
+		ExpectedStatusCode:   http.StatusNoContent,
+		ResponseObject:       nil,
+	})
+
+	return err
+}
+
+// ConfigureConsulAccess is to enable the Consul config access to the SecretStore via consul/config/access API
+// see the reference: https://www.vaultproject.io/api-docs/secret/consul#configure-access
+func (c *Client) ConfigureConsulAccess(secretStoreToken string, bootstrapACLToken string, registryHost string, registryPort int) error {
```
Comment (lenny-intel) [inline internal/pkg/vault/management.go:227]:
```suggestion
func (c *Client) ConfigureConsulAccess(secretStoreToken string, bootstrapACLToken string, consulHost string, consulPort int) error {
```

**Review by lenny-intel — CHANGES_REQUESTED** [review]
Code looks good, but need to change all usages of  `registry` to be `Consul`. I have flagged a bunch of them, but there are more.

**Comment by ItsRico** [comment]
Updated all references of registry to consul

File: internal/pkg/vault/management.go [hunk 8]
```diff
@@ -193,3 +193,57 @@ func (c *Client) CheckSecretEngineInstalled(token string, mountPoint string, eng
 
 	return false, nil
 }
+
+// CreateRole creates a secret store role that can be used to generate Consul tokens
```
Comment (lenny-intel) [inline internal/pkg/vault/management.go:197]:
```suggestion
// CreateRole creates a Consul role that can be used to generate Consul tokens
```

**Review by lenny-intel — APPROVED** [review]
LGTM, time to take out of draft?

**Review by lenny-intel — CHANGES_REQUESTED** [review]
One minor tweak needed

File: internal/pkg/vault/management.go (hunk 8 above)
Comment (ItsRico) [reply internal/pkg/vault/management.go]: updated

**Review by lenny-intel — APPROVED** [review]
LGTM

**Review by bnevis-i — APPROVED** [review]

**Pull request merged**